diffkemp/simpll/Logger.h

```diff
@@ -0,0 +1,113 @@
+        const auto &x = return_value;                                          \
+        LOG_KEEP(x);                                                           \
+        return x;                                                              \
+    } while (false)
+
+// Equivalent to LOG_KEEP_FORCE, additionally returing the given value
+#define RETURN_WITH_LOG_FORCE(return_value)                                    \
+    do {                                                                       \
+        const auto &x = return_value;                                          \
+        LOG_KEEP_FORCE(x);                                                     \
+        return x;                                                              \
+    } while (false)
+
+#define LOG(msg)                                                               \
+    DEBUG_WITH_TYPE(DEBUG_SIMPLL, dbgs() << getDebugIndent() << msg)
+#define LOG_NO_INDENT(msg) DEBUG_WITH_TYPE(DEBUG_SIMPLL, dbgs() << msg)
+#define LOG_VERBOSE(msg)                                                       \
+    DEBUG_WITH_TYPE(DEBUG_SIMPLL_VERBOSE, dbgs() << getDebugIndent() << msg)
+#define LOG_VERBOSE_EXTRA(msg)                                                 \
+    DEBUG_WITH_TYPE(DEBUG_SIMPLL_VERBOSE_EXTRA,                                \
+                    dbgs() << getDebugIndent() << msg)
+
+#define LOG_INDENT() DEBUG_WITH_TYPE(DEBUG_SIMPLL, increaseDebugIndentLevel())
+#define LOG_DEINDENT() DEBUG_WITH_TYPE(DEBUG_SIMPLL, decreaseDebugIndentLevel())
+
+#define WITHOUT_LOGGER_INDENT(actions)                                         \
+    do {                                                                       \
+        size_t indent;                                                         \
+        DEBUG_WITH_TYPE(LOGGER_BASE_LEVEL, indent = logger.storeIndent());     \
+        actions;                                                               \
+        DEBUG_WITH_TYPE(LOGGER_BASE_LEVEL, logger.loadIndent(indent));         \
+    } while (false)
+
+class Logger {
+  public:
+    Logger(){};
+    // prepare message for logging
+    llvm::raw_ostream &prep_log();
+    // prepare for logging messages within the context of the last prepared
+    // message
+    void prep_context();
+    // log a prepared message
+    void log(bool keep = true, const char *force_keep_type = nullptr);
+    // dump all messages from the buffer
+    void dump();
+    // reset the logger indent level to 0 and return the original value
+    size_t storeIndent();
+    // set the logger indent level to the given value
+    void loadIndent(size_t stored_level);
+
+  protected:
+    void increaseDebugIndentLevel();
+    void decreaseDebugIndentLevel();
+
+  private:
+    // current level of indentation within the logger
+    size_t level{0};
+    // debug message buffer (force-kept, indent level, message)
+    std::vector<std::tuple<bool, size_t, std::string>> buffer{};
```

Comment:
I suggest using a struct type whenever you need to document the meaning of the fields of a tuple, e.g. from seeing `if (!iter->force_kept)` (in Logger.cpp) you immediately know what the if condition is checking, unlike `if (!std::get<bool>(*iter))`.